protocol/consistencychecks.go

```diff
@@ -0,0 +1,265 @@
+	proofType = invalidProof
+
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	if verificationResult = cc.verifySTR(str); verificationResult != nil {
+		return
+	}
+
+	if proofType, verificationResult = cc.verifyAuthPath(uname, key,
+		ap, str); verificationResult != nil {
+		return
+	}
+
+	if verificationResult = cc.verifyReturnedPromise(tb, str, ap); verificationResult != nil {
+		return
+	}
+
+	if verificationResult = cc.verifyFulfilledPromise(str, ap); verificationResult != nil {
+		return
+	}
+
+	// re-assign new state for the client
+	if str.Epoch > cc.CurrentEpoch {
+		cc.CurrentEpoch++
+		cc.SavedSTR = str.Signature
+	}
+	return
+}
+
+func (cc *ConsistencyChecks) verifyDirectoryProofs(dfs *DirectoryProofs,
+	uname string, key []byte) (int, error) {
+	// TODO: fix me
+	return invalidProof, nil
+}
+
+func (cc *ConsistencyChecks) verifyAuthPath(uname string, key []byte,
+	ap *m.AuthenticationPath,
+	str *m.SignedTreeRoot) (int, error) {
+	proofType := proofOfAbsence
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return invalidProof, ErrorBadVRFProof
+	}
+
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) { // proof of inclusion */
+		// verify name-to-key binding
+		// key is nil when the user does lookup for the first time
+		if key != nil && !ap.VerifyBinding(key) {
+			return invalidProof, ErrorBadBinding
+		}
+		// verify commitment
+		if !ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+			return invalidProof, ErrorBadCommitment
+		}
+		proofType = proofOfInclusion
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return invalidProof, ErrorBadAuthPath
+	}
+
+	return proofType, nil
+}
+
+func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
+	// verify STR's signature
+	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	if (str.Epoch == cc.CurrentEpoch && bytes.Equal(cc.SavedSTR, str.Signature)) ||
+		(str.Epoch == cc.CurrentEpoch+1 && str.VerifyHashChain(cc.SavedSTR)) {
+		return nil
+	}
+	return ErrorBadSTR
+}
+
+// verifyReturnedPromise verifies the returned TB
+// included in the directory's response.
+func (cc *ConsistencyChecks) verifyReturnedPromise(tb *m.TemporaryBinding,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if cc.isUseTBs && tb != nil {
+		// verify TB's Signature
+		if !cc.signKey.Verify(tb.Serialize(str.Signature), tb.Signature) {
+			return ErrorBadSignature
+		}
+
+		// verify TB's VRF index
+		if !bytes.Equal(tb.Index, ap.LookupIndex) {
+			return ErrorBadIndex
+		}
+
+		cc.TBs = append(cc.TBs, tb)
+	}
+
+	return nil
+}
+
+// verifyFulfilledPromise verifies issued TBs was inserted
+// in the directory as promised.
+func (cc *ConsistencyChecks) verifyFulfilledPromise(str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if cc.isUseTBs {
+		backed := cc.TBs[:0]
+		for _, tb := range cc.TBs {
```

Comment:
> But my question is should the client "actively" verify all saved TBs of its contacts or just verify some TBs if it has the AP for that user?

I think the client shouldn't "actively" verify all saved TBs. Verifying if the TB promise was fulfilled in the next epoch can't be done without the corresponding AP for the particular private index, right? So the client can only verify individual TBs for which it has the AP (from a lookup or monitoring). In other words, the client should verify TBs on demand, i.e. if it does a lookup for alice, it should find all saved TBs for alice and check that they were included in the tree with her AP.